### DISCUSSION TITLE
autocommit`s effect

### DISCUSSION THREAD

**Discussion by panla** [sqlalchemy/sqlalchemy#6788] [Usage Questions]
if autocommit == False
   write need to commit and does read need to commit?
if autocommit == True
  do read and write not need to commit?

**Comment by zzzeek** [comment]
hi there -

the term "autocommit" is used in a few places in SQLAlchemy, although all are deprecated.

Can you please be more specific what APIs you are referring towards ?

**Reply by panla** [reply]
i get `session commit after read will be better` from other article, and i don`t know weather it is right.

**Reply by zzzeek** [reply]
a commit on a transaction is generally a little bit of an expense so i wouldn't seek to add this for a read only transaction if convenient.

the [commit-as-you-go](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#opening-and-closing-a-session) is likely the most straightforward pattern to use, where you call session.commit() only if you changed some data:

```py
with Session(engine) as sess:
    result = sess.execute(  .. query ...)

# done , you didn't write anything
``` 

commit as you go:

```py
with Session(engine) as sess:
    sess.add(my_object)
    sess.commit()  # commit as you go, if needed

# done , committed a row
```

**Reply by panla** [reply]
thanks
